packages/tldraw/src/lib/shapes/shared/useAsset.ts

```diff
@@ -1,31 +1,30 @@
-import { TLAssetId, useEditor, useValueDebounced } from '@tldraw/editor'
-import { useEffect, useState } from 'react'
+import { TLAssetId, debounce, useEditor } from '@tldraw/editor'
+import { useEffect, useRef, useState } from 'react'
 
 /** @internal */
 export function useAsset(assetId: TLAssetId | null, width: number) {
 	const editor = useEditor()
 	const [url, setUrl] = useState<string | null>(null)
 	const asset = assetId ? editor.getAsset(assetId) : null
+	const debouncedResolver = useRef(
+		debounce(async (screenScale: number) => {
+			const resolvedUrl = await editor.resolveAssetUrl(assetId, {
+				screenScale,
+			})
+			setUrl(resolvedUrl)
+		}, 500)
+	)
 
 	const shapeScale = asset && 'w' in asset.props ? width / asset.props.w : 1
 	// We debounce the zoom level to reduce the number of times we fetch a new image and,
 	// more importantly, to not cause zooming in and out to feel janky.
-	const debouncedScreenScale = useValueDebounced(
-		'zoom level',
-		() => editor.getZoomLevel() * shapeScale,
-		[editor, shapeScale],
-		500
-	)
+	const screenScale = editor.getZoomLevel() * shapeScale
 
 	useEffect(() => {
-		async function resolve() {
-			const resolvedUrl = await editor.resolveAssetUrl(assetId, {
-				screenScale: debouncedScreenScale,
-			})
-			setUrl(resolvedUrl)
-		}
-		resolve()
-	}, [assetId, debouncedScreenScale, editor])
+		const debouncedFn = debouncedResolver.current
+		debouncedFn(screenScale)
+		return () => debouncedFn.cancel()
```

Comment:
I think that this version with the `debounce` fn + `async` leaves a case where the component could unmount in the async gap after calling `resolveAssetUrl`. What if we move everything into the `useEffect`? (afaict, the timing of this is the same as a the debounce version?)

```suggestion
		let isCancelled = false
		const timer = setTimeout(() => {
			const resolvedUrl = await editor.resolveUrl(...)
			if (!isCancelled) setUrl(resolvedUrl)
		})
		return () => {
			clearTimeout(timer)
			isCancelled = true
		}
```

Reply:
sounds good to me! 👍